### ISSUE TITLE
Overlapping destination regions in copy commands

### ISSUE THREAD

**Issue by philiptaylor** [KhronosGroup/Vulkan-Docs#317]
If I do:

```
vkCmdCopyBuffer(
  srcBuffer = foo,
  dstBuffer = bar,
  regionCount = 2,
  pRegions = {
    { srcOffset = 0, dstOffset = 0, size = 1 },
    { srcOffset = 1, dstOffset = 0, size = 1 },
  }
);
```

what should the outcome be (i.e. what should end up being stored at byte 0 in the destination)? or should it be undefined behaviour?

The function description says:

> srcBuffer and dstBuffer can be the same buffer or alias the same memory, but the result is undefined if the copy regions overlap in memory.

and Valid Usage says:

> - The union of the source regions, and the union of the destination regions, specified by the elements of pRegions, must not overlap in memory

and the section intro says:

> - For non-sparse resources, the union of the source regions in a given buffer or image must not overlap the union of the destination regions in the same buffer or image. 
> - For sparse resources, the set of bytes used by all the source regions must not intersect the set of bytes used by all the destination regions. 

I interpret those as all trying to say the same thing (in variously ambiguous ways): if there is any byte of memory that is in a source region and also in a destination region, then it is undefined behaviour, for the same reason that e.g. `memcpy` forbids overlap.

But I don't think any of those sentences are talking about the case where one destination region overlaps another destination region. I expect that case should be forbidden. (Source regions overlapping other source regions should probably be permitted, though.)

(Alternatively, it could be defined to copy the regions sequentially in the order they were given in the `pRegions` array, with some suitable coherency behaviour for write-after-write between two regions. But if they were sequential, why not allow a region's source to overlap a previous region's destination? Therefore I assume the intention was that regions aren't sequential.)

Same applies to `vkCmdCopyImage`, `vkCmdCopyBufferToImage`, `vkCmdCopyImageToBuffer`, `vkCmdBlitImage`, `vkCmdResolveImage`.

**Comment by krOoze** [comment]
You are violating the union rule. Your example perhaps should be `srcOffset=1` and `srcOffset=2` to illustrate the real problem.

**Comment by philiptaylor** [comment]
It shouldn't be violating the union rule, since srcBuffer != dstBuffer, so the source-region-union and destination-region-union don't "overlap in memory" and are not "in the same buffer".

**Comment by krOoze** [comment]
^ Yeah I noticed that at last, hence the post deletion.

**Comment by philiptaylor** [comment]
Ah, GitHub seems to dynamically show new comments that arrive while you're writing one but doesn't show deletions until you reload the page, which is confusing. So I'll delete my comment, and this one, and nobody will ever know...

**Comment by jeffbolznv** [comment]
Doesn't this fall out of the general synchronization rules, that without a memory dependency between the two transfers they could overlap and produce a non-deterministic result?

It is, perhaps, inconsistent to forbid overlapping source and destination within a command but not overlapping solely within the source or destination.

**Comment by krOoze** [comment]
@jeffbolznv perhaps, but you wouldn't expect self-contained single command to couse synchronization issues. Also there's even no mechanism to synchronize them. Also, as you say, it would be inconsistent to ban one and not the other.

**Comment by philiptaylor** [comment]
> Doesn't this fall out of the general synchronization rules, that without a memory dependency between the two transfers they could overlap and produce a non-deterministic result?

Hmm, maybe. I think the main problem with that argument is that memory dependencies are defined in terms of memory accesses performed by commands, but nothing defines what memory accesses are actually performed by commands.

I think `vkCmdClearColorImage` (also `vkCmdClearDepthStencilImage`, `vkCmdClearAttachments`) has a similar problem but demonstrates it more clearly. It's just defined to clear one or more subranges - does that mean it's performing a separate write for each subrange in sequence (in which case overlapping ranges would cause a write-after-write hazard and probably undefined behaviour), or that it's performing the minimum number of writes that will result in all the subranges becoming cleared (in which case overlaps are fine)?

Without some knowledge of the internal accesses performed by each command, I don't think you can understand them properly using the normal memory dependency rules. Either those accesses should be defined clearly, or the commands should have explicit validity requirements that prevent those hazardous cases from ever being an issue.

(The copy commands aren't as ambiguous as the clear ones (it's hard to see how they could do anything other than read and write each region separately), but their accesses are still not really defined at all.)

> you wouldn't expect self-contained single command to couse synchronization issues.

I think you sometimes would, e.g. two shader invocations in a single draw call can't safely access the same memory because of synchronisation, unless it's `Coherent` or atomic or a colour/depth attachment etc. But for draw commands the spec goes into some detail about what is allowed, and it's fairly obvious what memory accesses are happening, whereas the transfer commands don't give any detail like that, so it's harder to know what to expect from them.

**Comment by jeffbolznv** [comment]
> Hmm, maybe. I think the main problem with that argument is that memory dependencies are defined in terms of memory accesses performed by commands, but nothing defines what memory accesses are actually performed by commands.

I disagree. The spec is clear about what memory is read and written by these commands and which pixels/bytes of the source correspond to which pixels/bytes of the destination, and in cases where it is not totally obvious (e.g. vkCmdCopyBuffer is obvious, vkCmdCopyImageToBuffer is not totally obvious) we have code/formulae to specify it. And those accesses are performed by the transfer stage of the pipeline. Any further details of **how** the copy is performed are left to the implementation to decide. For example, an implementation could launch a bunch of compute shader invocations to do all the copy regions in parallel, and another implementation could have a dedicated DMA engine that does the copies in order (or it could overlap the copies).

The spec is also fairly clear that work can execute out of order or overlap unless there is an explicit dependency. Maybe it doesn't explicitly say that copies generated by a single command may execute out of order or overlap, but the whole execution/synchronization model of Vulkan starts from a premise of "unordered unless requested by the app", in contrast to GL's execution model of "as if each operation executed immediately, in order".

**Comment by jeffbolznv** [comment]
@philiptaylor, would you be satisfied to resolve this by updating the "Common Operations" section to also forbid overlapping destination regions (which I think is what you originally wanted)?

**Comment by philiptaylor** [comment]
Yes, I think that would make me happy enough.

It wouldn't resolve the similar problem with `vkCmdClearColorImage`, `vkCmdClearDepthStencilImage` - if they have overlapping `pRanges`, is that okay (each pixel will only be cleared once) or is it an illegal write-after-write hazard? And similarly for `vkCmdClearAttachments` with overlapping `pRects`, or with `pAttachments` containing one attachment multiple times (potentially with different clear values) (or potentially two different attachments that are aliased). Maybe that should be raised as a separate issue?

**Comment by oddhack** [comment]
This should be done in the 1.0.65 spec update.
